oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java

```
@@ -0,0 +1,82 @@
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class MongoClusterListener implements ClusterListener {
+
+    // Sometimes we need to wait a few seconds in case the connection was just created, the listener
+    // didn't have time to receive the description from the cluster. This latch is used to check
+    // if the connection was properly initialized.
+    private final CountDownLatch LATCH;
+    private final int LATCH_AWAIT_TIMEOUT = 15;
+    private boolean replicaSet = false;
+    private ServerAddress serverAddress;
+    private ServerAddress primaryAddress;
+
+    public MongoClusterListener() {
+        LATCH = new CountDownLatch(1);
+    }
+
+    @Nullable
+    public ServerAddress getServerAddress() {
+        waitForLatch();
+        return serverAddress;
+    }
+
+    @Nullable
+    public ServerAddress getPrimaryAddress() {
+        waitForLatch();
+        return primaryAddress;
+    }
+
+    public boolean isReplicaSet() {
+        waitForLatch();
+        return replicaSet;
+    }
+
+    @Override
+    public void clusterDescriptionChanged(final ClusterDescriptionChangedEvent event) {
+        for (ServerDescription sd : event.getNewDescription().getServerDescriptions()) {
+            if (sd.getState() == ServerConnectionState.CONNECTED) {
+                serverAddress = sd.getAddress();
+                primaryAddress = new ServerAddress(sd.getPrimary());
+                if (sd.isReplicaSetMember()) {
+                    // Can't assign directly the result of the function because in some cases the cluster
+                    // type is UNKNOWN, mainly when the cluster is changing it's PRIMARY.
+                    replicaSet = true;
+                }
+                LATCH.countDown();
+            }
+        }
+    }
+
+    private void waitForLatch() {
+        try {
+            LATCH.await(LATCH_AWAIT_TIMEOUT, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {}
```

I would add a `log.error` msg here instead of swallowing the exception.